Sharding relations test are flaky

Gu1nness opened the pull request:
## Issue

 * Test fails 90% of the time.
 * The real issue here is an unprotected call to `charm.client_provider.update_app_relation_data` which would create the password only for an invalid relation (integrating a shard as database).
 * This call is run in every status check.

## Solution

 * Protect the call by running some sanity checks.
 * Update all tests that started failing due to this (needed to add some monkey patching for all methods used in this checks.

## Bug fixes

 * The `charm.db_initialised` function was buggy: setting `charm.db_initialised = False` would still return `charm.db_initialised == True`. This is now fixed in a retro compatible way to ensure that nothing breaks with upgrades.
 * This fix needs to be ported to mongodb k8s as well.

Gu1nness committed:
wip: try removing pin

Gu1nness committed:
fix: Do not update relation data if we cannot pass hook checks

Gu1nness committed:
fix(db_initialised): Fix bug in db_initialised

The function was only checking for the existence of the field, not its
value. This meant that `{"db_initialised": "false"}` would lead to
`charm.db_initialised == True`.
Also adapts all tests for that and missing mocks due to former commit.

Gu1nness committed:
fix: formatting

Gu1nness committed:
fix: more stable version

Gu1nness committed:
fix: ensure casing

Gu1nness committed:
fix: lowercase

MiaAltieri commented:
Note for readers test failing will be addressed #473

Mehdi-Bendriss commented on lib/charms/mongodb/v1/mongodb_provider.py:
```diff
@@ -76,8 +76,8 @@ def __init__(self, charm: CharmBase, substrate="k8s", relation_name: str = REL_N
 
-    def pass_hook_checks(self, event: EventBase) -> bool:
-        """Runs the pre-hooks checks for MongoDBProvider, returns True if all pass."""
+    def sanity_hook_cheks(self) -> bool:
```
1. Could you rename this method with the `is_` prefix? it is unclear from the name what this check does. 
2. Why don't we extend the existing `pass_hook_checks` and call it everywhere instead of a second method?

Gu1nness replied:
The `pass_hook_checks` has two parts:
 * The first one which I extracted to the `sanity_hook_checks` which is event independent
 * The second part is event dependent and can defer the event.
 
 In this case, we don't want to defer the `update_status` event, we just want to protect the call.
 That's why I decided to split the hook checks between the sanity ones which always have to be run, and the ones that can defer the event.
The existing `pass_hook_checks` hasn't changed it what it does, I just went for reusability :)

Gu1nness replied:
> 1. Could you rename this method with the `is_` prefix? it is unclear from the name what this check does.

I think I'll go to `pass_sanity_hook_checks` which is a method name that we already have in other libraries, this way we'll keep some uniformity here.

Mehdi-Bendriss reviewed (commented):
Thanks Neha - left a clarification point.

Gu1nness committed:
fix: rename

MiaAltieri commented on tests/integration/sharding_tests/test_sharding_relations.py:
```diff
@@ -55,7 +55,6 @@ async def test_build_and_deploy(
     await ops_test.model.deploy(
         MONGOS_APP_NAME,
         channel="6/edge",
-        revision=3,
```
ty

Gu1nness committed:
fix: Enfore libjuju version

With the last libjuju version (3.5.2), there is a bug when trying to
deploy charms with custom storage
(https://github.com/juju/python-libjuju/issues/1093)
This pins the libjuju version until it's fixed
